src/main/java/com/minecolonies/entity/ai/citizen/builder/EntityAIStructureBuilder.java

```diff
@@ -344,35 +380,64 @@ private BlockPos getPositionInDirection(EnumFacing facing, int distance)
     @NotNull
     private BlockPos getFloor(@NotNull BlockPos position)
     {
+        final BlockPos floor = getFloor(position, 0);
+        if (floor == null)
+        {
+            return position;
+        }
+        return floor;
+    }
+
+    /**
+     * Calculates the floor level
+     *
+     * @param position input position
+     * @param depth    the iteration depth
+     * @return returns BlockPos position with air above
+     */
+    @Nullable
+    private BlockPos getFloor(@NotNull BlockPos position, int depth)
+    {
+        if (depth > 50)
+        {
+            return null;
+        }
         //If the position is floating in Air go downwards
         if (!EntityUtils.solidOrLiquid(world, position))
         {
-            return getFloor(position.down());
+            return getFloor(position.down(), depth + 1);
         }
         //If there is no air above the block go upwards
         if (!EntityUtils.solidOrLiquid(world, position.up()))
         {
             return position;
         }
-        return getFloor(position.up());
+        return getFloor(position.up(), depth + 1);
     }
 
     private AIState clearStep()
     {
         WorkOrderBuild wo = job.getWorkOrder();
 
+        if (job.getStructure() == null)
+        {
+            //fix for bad structures
+            job.complete();
+        }
+
         if (wo.isCleared())
         {
             return AIState.BUILDER_STRUCTURE_STEP;
         }
 
-        BlockPos coordinates = job.getSchematic().getBlockPosition();
-        Block worldBlock = world.getBlockState(coordinates).getBlock();
+        BlockPos coordinates = job.getStructure().getBlockPosition();
+        IBlockState worldBlockState = world.getBlockState(coordinates);
+        Block worldBlock = worldBlockState.getBlock();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'worldBlock' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)